File: Programmers/난이도/Level2/게임 맵 최단거리/code/solution.js
function solution(maps = [[]]) {
  let n = maps.length;
  let m = maps[0].length;

  let queue = [];
  let diff = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];

  queue.push([0, 0]);

  // BFS
  while (queue.length > 0) {
    let v = queue.splice(0, 1);
    let x = v[0][0];
    let y = v[0][1];

    for (let d of diff) {
      let nx = x + d[0];
      let ny = y + d[1];

      // 다음 갈 좌표가 유효한 영역내에 있는지 확인
      if (!(0 <= nx && nx < n) || !(0 <= ny && ny < m)) {
        continue;
      }

      // 이미 방문한 좌표인지 혹은 벽인지 확인
      if (maps[nx][ny] != 1) {
        continue;
      }

      // 다음 갈 좌표가 목적지 이면 함수 종료 및 경로 길이 반환
      if (nx == n - 1 && ny == m - 1) {
        return maps[x][y] + 1;
      }

      // 경로 길이 갱신
      maps[nx][ny] = maps[x][y] + 1;

      // 다음 좌표 Queue에 삽입
      queue.push([nx, ny]);
    }
  }

  return -1;
}

console.log(
  solution([
    [1, 0, 1, 1, 1],
    [1, 0, 1, 0, 1],
    [1, 0, 1, 1, 1],
    [1, 1, 1, 0, 1],
    [0, 0, 0, 0, 1],
  ])
);
console.log(
  solution([
    [1, 0, 1, 1, 1],
    [1, 0, 1, 0, 1],
    [1, 0, 1, 1, 1],
    [1, 1, 1, 0, 0],
    [0, 0, 0, 0, 1],
  ])
);

File: Programmers/난이도/Level2/게임 맵 최단거리/code/solution.py
from collections import deque


def solution(maps=[]):
    n = len(maps)
    m = len(maps[0])
    queue = deque([[0, 0]])

    # 상하좌우
    diff = [[-1, 0], [1, 0], [0, -1], [0, 1]]

    # BFS
    while queue:
        v = queue.popleft()
        x = v[0]
        y = v[1]

        for d in diff:
            nx = x + d[0]
            ny = y + d[1]

            # 갈 수 있는 좌표인지 확인
            if not (nx >= 0 and nx < n) or not (ny >= 0 and ny < m):
                continue

            # 이미 방문한 좌표인지 혹은 벽인지 확인
            if maps[nx][ny] != 1:
                continue

            # 이동할 점이 목표 지점이라면 경로의 길이 반환
            if nx == n - 1 and ny == m - 1:
                return maps[x][y] + 1

            # 방문할 좌표 추가
            queue.append([nx, ny])

            # 방문할 좌표에 대해 경로 길이 갱신
            maps[nx][ny] = maps[x][y] + 1

    return -1


if __name__ == "__main__":
    print(
        solution(
            [
                [1, 0, 1, 1, 1],
                [1, 0, 1, 0, 1],
                [1, 0, 1, 1, 1],
                [1, 1, 1, 0, 1],
                [0, 0, 0, 0, 1],
            ]
        )
    )
    print(
        solution(
            [
                [1, 0, 1, 1, 1],
                [1, 0, 1, 0, 1],
                [1, 0, 1, 1, 1],
                [1, 1, 1, 0, 0],
                [0, 0, 0, 0, 1],
            ]
        )
    )
